llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -9328,10 +9328,25 @@ AArch64TargetLowering::LowerCall(CallLoweringInfo &CLI,
       }
       unsigned LocMemOffset = VA.getLocMemOffset();
       int32_t Offset = LocMemOffset + BEAlign;
-      SDValue PtrOff = DAG.getIntPtrConstant(Offset, DL);
-      PtrOff = DAG.getNode(ISD::ADD, DL, PtrVT, StackPtr, PtrOff);
 
       if (IsTailCall) {
+        // When the frame pointer is perfectly aligned for the tail call and the
+        // same stack argument is passed down, omit storing it if is immutable
+        // and already in the right offset.
+        if (FPDiff == 0) {
+          if (auto *LoadNode = dyn_cast<LoadSDNode>(Arg)) {
+            if (auto *FINode =
+                    dyn_cast<FrameIndexSDNode>(LoadNode->getBasePtr())) {
+              MachineFrameInfo &MFI = MF.getFrameInfo();
+              int FI = FINode->getIndex();
+              if (LoadNode->getMemoryVT() == VA.getValVT() &&
```

Comment:
I think you should add a test for `i1` args, whose memory type is different than their in-registers type.

Reply:
Also, probably needs a couple tests for `zeroext`/`sext` args to make sure we do the right thing for smaller-than-32-bits args. In particular, we should make sure that when caller/callee don't agree on sign extension, that we still emit the caller-required part of that, and when they do agree we may omit it.